nupic/research/frameworks/greedy_infomax/models/BilinearInfo.py

```diff
@@ -0,0 +1,198 @@
+            ztwk = (
+                self.W_k[k - 1]  # 1x1 convolution
+                .forward(z[:, :, (k + skip_step) :, :])  # Bx, C , H , W
+                .permute(2, 3, 0, 1)  # H, W, Bx, C
+                .contiguous()
+            )  # y, x, b, c
+
+            # Take random samples from z_{t+k} W_k as contrastive samples
+            ztwk_shuf = ztwk.view(
+                ztwk.shape[0] * ztwk.shape[1] * ztwk.shape[2], ztwk.shape[3]
+            )  # y * x * batch, c
+            rand_index = torch.randint(
+                ztwk_shuf.shape[0],  # y *  x * batch
+                (ztwk_shuf.shape[0] * self.negative_samples, 1),
+                dtype=torch.long,
+                device=ztwk_shuf.device,
+            )
+            rand_index = rand_index.repeat(1, ztwk_shuf.shape[1])
+            ztwk_shuf = torch.gather(
+                ztwk_shuf, dim=0, index=rand_index, out=None
+            )  # y * x * b * n, c
+            ztwk_shuf = ztwk_shuf.view(
+                ztwk.shape[0],
+                ztwk.shape[1],
+                ztwk.shape[2],
+                self.negative_samples,
+                ztwk.shape[3],
+            ).permute(
+                0, 1, 2, 4, 3
+            )  # y, x, b, c, n
+
+            # Multiply ztwk and context for full bilinear model
+            context = (
+                c[:, :, : -(k + skip_step), :].permute(2, 3, 0, 1).unsqueeze(-2)
+            )  # y, x, b, 1, c
+
+            log_fk_main = torch.matmul(context, ztwk.unsqueeze(-1)).squeeze(
+                -2
+            )  # y, x, b, 1
+
+            log_fk_shuf = torch.matmul(context, ztwk_shuf).squeeze(-2)  # y, x, b, n
+
+            # Stack negative samples below positive samples
+            log_fk = torch.cat((log_fk_main, log_fk_shuf), 3)  # y, x, b, 1+n
+            log_fk = log_fk.permute(2, 3, 0, 1)  # b, 1+n, y, x
+
+            # Positive samples are at index 0
+            true_fk = torch.zeros(
+                (batch_size, log_fk.shape[-2], log_fk.shape[-1]),
+                dtype=torch.long,
+                device=log_fk.device,
+            )  # b, y, x
+
+            # #append results to list
+            log_f_list.append(log_fk)
+            true_f_list.append(true_fk)
+
+        return log_f_list, true_f_list
+
+
+class SparseBilinearInfo(BilinearInfo):
+    def __init__(
+        self,
+        in_channels,
+        out_channels,
+        negative_samples=16,
+        k_predictions=5,
+        sparse_weights_class=SparseWeights2d,
+        sparsity=0.1,
+    ):
+        super(SparseBilinearInfo, self).__init__(
+            in_channels, out_channels, negative_samples, k_predictions
+        )
+        if sparsity and sparsity > 0.3:
+            for i in range(len(self.W_k)):
+                self.W_k[i] = sparse_weights_class(self.W_k[i], sparsity=sparsity)
+
+
+class VDropSparseBilinearInfo(BilinearInfo):
+    def __init__(
+        self,
+        in_channels,
```

Comment:
Should this really be a separate class, or should `BilinearInfo` / `SparseBilinearInfo` just have a regularization option that allows for VDrop?

Reply:
In other models, VDrop versions are explicitly created as classes. In this case, it's necessary to introduce the central_data parameter which is specific to VDrop.